### PR TITLE
RevDiff custom difftool

### LINKED ISSUE #7716: User selectable merge and difftool
## Feature description 
All merge and difftools are not best in every situation. It can be good to be able to select a custom tool in some situations.

Git command line allows selecting default and user configured tools:
```git mergetool --tool=kdiff3```

The available tools can be listed:
```git mergetool --tool-help```

This commands also lists the .cmd for the user defined tools. The listing takes a couple of seconds (likely tests paths for git built-in tools).

The tools can be made available as a submenu:

mergetool is only available in FormResolveConflicts
 * Upper right button starts mergetool for selected file only with GE manipulations, requiring that mergetool.cmd is configured. #7655 changes the requirement to only make the extra processing where .cmd is required for those commands.
(I proposed to remove some of this extra handling in #7550.)
 * Second button starts git-mergetool without any GE manipulation, for all files.

For both these buttons, the available tools ban be made available as a button context menu

image

difftool is available in many situations.
The extra difftool can be made as a submenu in each place a difftool is available.
RevisionDiff is the obvious place to add this.

image

Open issues:
 * To list available tools requires seconds (at work, the Symantec virus delays Git operations like git-status 10 times). This may have to be done async when the form is opened.
 * Git lists some toolnames many times, for instance all mergetools together with difftools. It should be OK to list unique names.
 * For mergetools .cmd could be used in the GE extra handling, but it is not clear what .cmd that Git will use, so this part can be skipped (#7655)

## Environment

- Git Extensions 3.3.0
- Build 1e98d521d3c99de6a7afd33a01dcbd0e62358b95
- Git 2.25.0.windows.1
- Microsoft Windows NT 6.2.9200.0
- .NET Framework 4.8.4075.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by gerhardol**
Fixes #7716

Based on #8192 

## Proposed changes

Add the diff tools known to Git in a submenu in RevDiff
The list is collected asynch when the form is first created

### After

image

## Test methodology 

Added a test

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
Added a fish-need-bike test, similar to most existing tests
A slightly more comprehensive test involving Git could be implemented, but I fail now (and do not want to spend more time on this).

**Comment by gerhardol** [comment]
> A comment or `enum` regarding `mode` in `ParseCustomDiffMergeTool` would be helpful.

Added a few comments.
A regexp approach could be used when parsing too (see the issue for an example).
 
> I'd prefer a central setting, e.g. in the toolbar and/or the context menus, in favor of having a single (or few) place(s) with a submenu.
> Every place to open a difftool is of interest for me. (E.g. the history form leaves much room for improvements regarding external diffs.)

This is enough for me (a splitbutton for merging would be good too).
If I want it in some other situation, I can go and change the tool.
To change the tool would be an extra step that I would like to avoid.

File: GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs [hunk 1]
```diff
@@ -171,6 +173,13 @@ private void InitializeComponent()
             this.openWithDifftoolToolStripMenuItem.Text = "Open with difftool";
             this.openWithDifftoolToolStripMenuItem.DropDownOpening += new System.EventHandler(this.openWithDifftoolToolStripMenuItem_DropDownOpening);
             // 
+            // openWithCustomDifftoolToolStripMenuItem
+            // 
+            this.openWithCustomDifftoolToolStripMenuItem.Image = global::GitUI.Properties.Images.Diff;
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:178]: The intention here was that the submenu implicated "something more", but I will remove it

File: GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs [hunk 2]
```diff
@@ -171,6 +173,13 @@ private void InitializeComponent()
...
+            this.openWithCustomDifftoolToolStripMenuItem.Name = "openWithCustomDifftoolToolStripMenuItem";
+            this.openWithCustomDifftoolToolStripMenuItem.Size = new System.Drawing.Size(296, 22);
+            this.openWithCustomDifftoolToolStripMenuItem.Text = "Open with custom difftool";
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:181]: OK

**Comment by gerhardol** [comment]
> There's no need to hide easter eggs for me, e.g. in the nevertheless very helpful comment. 😉

@mstv do you always want a mention for issues you are already reviewing?

Yes settings is a little slow and custom tools are not visible (I cannot use _tortoisemerge_ but have to use a custom _tortoisemerge2_  too - should investigate). 
But it is still more often I select the commits I want to compare with a specific tool then I need to switch the default difftool in the other than "F3" diffs.

A fast switcher could be implemented in addition of course. 
There is plenty of time to change before 4.0 will ship anyway.

mergetool is only invoked from one form, so a splitbutton there is natural, no need for a fast switcher. (not in this PR anyway)

**Comment by RussKie** [comment]
I'm itching to look at why the settings dialog is so slow, it annoys the
hell out of me

File: GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:181]: First -> Second...

File: GitCommands/Git/GitCommandHelpers.cs [hunk 3]
```diff
@@ -1018,6 +1018,66 @@ public static IReadOnlyList<GitItemStatus> GetSkipWorktreeFilesFromString(string
...
+            // Simple parsing of the textual output opposite to porcelain format
+            // Explanation of the modes:
+            //  0: Initial mode (to detect first line, direct transit to next mode)
+            //  1: Built in tools, in the path (first is header, ended by a newline)
+            //  2: Newlines separating tools
+            //  3: Built in tools, in the path (first is header, ended by a newline)
+            //  4: Newlines separating tools
+            //  5: Unavailable tools, skipped
+            //  6: End note, skipped
+            //  5: Newlines
```
Comment (mstv) [inline GitCommands/Git/GitCommandHelpers.cs:1039]: > > There's no need to hide easter eggs for me, e.g. in the nevertheless very helpful comment. 😉

> Do you always want a mention for issues you are already reviewing?

No, I meant just the minor inconsisties in the comment above. I didn't want to request a change.

Afterwards, I had a look at the screenshots again.

**Comment by gerhardol** [comment]
@mstv 
I compared to the Git code for generating the list some more and adapted the parsing. It is simpler too (only one mode). Will not work if this command is translated.
There can be other approaches to the parsing, like excluding the group starting with "The following tools are valid, but not currently available:", to limit damage from output changes.
But there should be a parsable output.
If I get response from Git maintainers, I may submit a patch:
https://github.com/git/git/pull/471#issuecomment-660205096

File: GitCommands/Git/GitCommandHelpers.cs [hunk 4]
```diff
@@ -1018,6 +1018,54 @@ public static IReadOnlyList<GitItemStatus> GetSkipWorktreeFilesFromString(string
...
+                // Ignore tools that must run in a terminal
+                string[] ignoredTools = { "vimdiff", "vimdiff2", "vimdiff3" };
+                var toolName = tool[0];
+                if (!string.IsNullOrWhiteSpace(toolName) && !tools.Contains(toolName) && !ignoredTools.Contains(toolName))
+                {
+                    tools.Add(toolName);
+                }
+            }
+
+            return tools.OrderBy(i => i[0]).ToList();
```
Comment (mstv) [inline GitCommands/Git/GitCommandHelpers.cs:1066]: Why sort only by the first character? `i` is a `string` IIUIC. So `.Sort()` should suffice.

Comment (gerhardol) [reply GitCommands/Git/GitCommandHelpers.cs]: .Sort() on a list requires comparator, but OrderBy(i => i) works

**Comment by gerhardol** [comment]
> The list seems to be not updated after closing FormSettings. I have waited more than the ~10 seconds it takes to fill the list after startup.

That was not implemented, not considered needed before. But a few lines of code, so I updated

**Comment by gerhardol** [comment]
Squashed the approved commits, rebased

I tried to motivate a patch to Git to give a safer and simpler output but do not believe I can motivate such a change in Git.
The response would likely be to implement the parsing with the raw Git commands instead (but you need to access internal Git structures to get tools that are in the Git path, so those will be excluded.).
Added a few comments about that.

I also slightly changed the parsing to risk show invalid tools rather than excluding valid tools if Git output format changes. (This has not changed since the --tool-help option was added.)
Slightly simpler parsing, so I hope @mstv will approve

If approved by mstv, I would like to merge this in a couple of days

**Pull request merged**